pkg/test/framework/components/apps/kube.go

```diff
@@ -608,16 +611,13 @@ func (a *kubeApp) Call(e AppEndpoint, opts AppCallOptions) ([]*echo.ParsedRespon
 	}
 
 	if len(resp) != 1 {
-		return nil, fmt.Errorf("unexpected number of responses: %d", len(resp))
+		return resp, fmt.Errorf("unexpected number of responses: %d", len(resp))
```

Comment:
I don't think returning the response here is correct. If the result is nil, then depending on the result return value seems wrong.

Reply:
I don't see what is wrong is wrong with returning both a value and an error? The alternative is to remove all error handling, which would push it to the caller side for most tests which would be annoying, or make a new function.

I need to validate that we did get a 502 error in this case. I am sure there are other cases where we want to validate that an error occurred as well (I know Liam is working on one), and parsing the err message is much worse I think.

I think it is much simpler to do this then create another function that has to be kept in sync with native.

Reply:
It is confusing and complicates handling code. For example, what happens when nil and err returned? it can happen in this method. It also goes counter to how we do error handling in the rest of the project. The common idiom is, once an error is returned, you cannot depend on the result.

An error from the Call() implies that there was something seriously wrong with making the call itself. If the error checks in the method is not of this category (i.e. they are not checking for some serious failure), then that shouldn't be a failure, and it is the responsibility of the caller to validate the result. If the validations are quite common, then it makes sense to make a utility method to perform validations, either on the returned item, as a static helper, or a wrapper that calls and validates.

Reply:
Ok, sounds reasonable. Do you think it makes sense to move this out to CallOrFail? Maybe rename it to something more like CallOrFailIfNotValidResponse except with a better name?

Reply:
I think whether Call should fail depends on use cases. As @howardjohn mentioned, different tests can have different criteria,

- 503 is an expected response for some negative test.
- call even fails at mtls handshake level for some security related test.
- More common cases, where we expect A to B will succeed, existing CallOrFail sematics can do a good job.

IMHO, we shouldn't conflate these two senarios into a single methods and make API harder to follow. We should add another method, something like TryCall() and return detailed error information and hand it over to test writers themselves to decide.

Once we have more tests to determine what are common patterns, we may then be able to provide some syntax sugar at framework utilities function.

Reply:
There are 2 types of errors that are fairly distinct here:

1. Errors making the grpc call itself (e.g. connection failures, timeouts, un-parsable response)
2. Successful grpc requests that returned parsable error responses

I believe that the code we have today (correct me if I'm wrong) will return an error for the first condition only. For the second, it will return a nil error along with the parsed responses. That, I believe, is the correct behavior and we should preserve it.  If that's not the behavior today, we should change it.

Reply:
The current code returns and error for case 1 and 2. This PR changes it to return an error only for one.

Reply:
So when a TCP request is rejected due to authorization, the error message is EOF. What is the error message if gprc fails in case 1 (and/or when the callee service is down)? I assume you guys are talking about this check https://github.com/istio/istio/blob/master/pkg/test/framework/components/apps/kube.go#L663

Reply:
Yeah I run into the same issue @pitlv2109 see https://github.com/istio/istio/pull/13440/files#diff-94dfa3425281cca12d7d39240c52c7c2R130. So we really need to support all degrees of error handling

Reply:
Sorry I don't get it. Are you saying if gRPC fails, it will return EOF as well?

Reply:
Call will return an error if:
* the endpoint given is invalid (should never happen)
* ForwardEcho returns an error, which happens when the grpc Invoke function returns an error. So this will be an error directly from grpc. This will give an error if we hit the BlackholeCluster (I believe this is what happens when the request is unauthorized?).

ValidatedCall will return an error if
* Call fails
* Validations fail: host, non 2xx, etc

Reply:
I understand that :D but my question is mostly what error message will Call gets if ForwardEcho fails rather than due to TCP authorization. I wonder if another case it will have EOF as well, so there will be conflicts and RBAC needs to find another way to distinguish them. We can test this by shutting down a testing service and make the call directly from the container.